clients/cli/src/prover_runtime.rs

```diff
@@ -0,0 +1,446 @@
+) -> JoinHandle<()> {
+    tokio::spawn(async move {
+        let mut next_worker = 0;
+        loop {
+            tokio::select! {
+                Some(task) = task_receiver.recv() => {
+                    let target = next_worker % worker_senders.len();
+                    if let Err(_e) = worker_senders[target].send(task).await {
+                        // eprintln!("Dispatcher: failed to send task to worker {}: {}", target, e);
+                        // TODO:
+                    }
+                    next_worker += 1;
+                }
+
+                _ = shutdown.recv() => {
+                    break;
+                }
+            }
+        }
+    })
+}
+
+/// Spawns a set of worker tasks that receive tasks and send prover events.
+///
+/// # Arguments
+/// * `num_workers` - The number of worker tasks to spawn.
+/// * `results_sender` - The channel to emit results (task and proof).
+/// * `prover_event_sender` - The channel to send prover events to the main thread.
+///
+/// # Returns
+/// A tuple containing:
+/// * A vector of `Sender<Task>` for each worker, allowing tasks to be sent to them.
+/// * A vector of `JoinHandle<()>` for each worker, allowing the main thread to await their completion.
+pub fn start_workers(
+    num_workers: usize,
+    results_sender: mpsc::Sender<(Task, Proof)>,
+    event_sender: mpsc::Sender<WorkerEvent>,
+    shutdown: broadcast::Receiver<()>,
+) -> (Vec<mpsc::Sender<Task>>, Vec<JoinHandle<()>>) {
+    let mut senders = Vec::with_capacity(num_workers);
+    let mut handles = Vec::with_capacity(num_workers);
+
+    for worker_id in 0..num_workers {
+        let (task_sender, mut task_receiver) = tokio::sync::mpsc::channel::<Task>(8);
+        let prover_event_sender = event_sender.clone();
+        let results_sender = results_sender.clone();
+        let mut shutdown = shutdown.resubscribe(); // Clone the receiver for each worker
+        let handle = tokio::spawn(async move {
+            while let Some(task) = task_receiver.recv().await {
+                // Check for shutdown signal
+                tokio::select! {
+                    _ = shutdown.recv() => {
+                        let message = format!("Worker {} received shutdown signal", worker_id);
+                        let _ = prover_event_sender
+                            .send(WorkerEvent::Prover {
+                                worker_id,
+                                data: message,
+                            })
+                            .await;
+                        break; // Exit the loop on shutdown signal
+                    }
+                    _ = tokio::time::sleep(Duration::from_millis(100)) => {
+                        // Continue processing the task
+                        let stwo_prover =
+                           crate::prover::get_default_stwo_prover().expect("Failed to create prover");
+                        match authenticated_proving(&task, stwo_prover).await {
```

Comment:
The `stwo_prover` might be a little memory expensive to be cloning it around like this. Is there a reason not just to construct it within the `authenticated_proving` function?

Reply:
Good suggestion